sdk/search/Azure.Search.Documents/src/AzureSearchDocumentsEventSource.cs

```diff
@@ -15,35 +18,94 @@ internal sealed class AzureSearchDocumentsEventSource : EventSource
-        public void BatchActionCountUpdated(string endPoint, int oldBatchCount, int newBatchCount)
+        [Event(BatchActionCountUpdatedEvent, Level = EventLevel.Warning, Message = "{0} at {1} has updated the starting batch action count from {2} to {3}.")]
+        public void BatchActionCountUpdated(string publisherType, string endPoint, int oldBatchCount, int newBatchCount)
         {
-            WriteEvent(BatchActionCountUpdatedEvent, endPoint, oldBatchCount, newBatchCount);
+            WriteEvent(BatchActionCountUpdatedEvent, publisherType, endPoint, oldBatchCount, newBatchCount);
+        }
+
+        [Event(PublisherDisposedWithPendingActionsEvent, Level = EventLevel.Error, Message = "{0}: {1} unsent indexing actions at {2}.")]
+        public void PublisherDisposedWithPendingActions(string componentType, string endPoint, int indexingActionsCount)
+        {
+            WriteEvent(PublisherDisposedWithPendingActionsEvent, componentType, indexingActionsCount, endPoint);
+        }
+
+        [Event(DocumentsPublishedEvent, Level = EventLevel.Verbose, Message = "{0}: publishing documents. Flush = {2}")]
+        public void PublishingDocuments(string publisherType, bool flush)
+        {
+            WriteEvent(DocumentsPublishedEvent, publisherType, flush);
+        }
+
+        [Event(ActionNotificationEventHandlerExceptionThrownEvent, Level = EventLevel.Error, Message = "{0}: exception thrown at {1}. Action = {2}. Exception: {3}")]
+        public void ActionNotificationEventHandlerExceptionThrown(string componentType, string endPoint, string action, string e)
+        {
+            WriteEvent(ActionNotificationEventHandlerExceptionThrownEvent, componentType, endPoint, action, e);
+        }
+
+        [NonEvent]
+        public void ActionNotificationEventHandlerExceptionThrown(string componentType, string endPoint, string action, Exception e)
+        {
+            if (IsEnabled(EventLevel.Error, EventKeywords.All))
+            {
+                ActionNotificationEventHandlerExceptionThrown(componentType, endPoint, action, FormatException(e));
+            }
+        }
+
+        [NonEvent]
+        private static string FormatException(Exception ex)
+        {
+            StringBuilder sb = new();
+            bool nest = false;
+
+            do
+            {
+                if (nest)
+                {
+                    // Format how Exception.ToString() would.
+                    sb.AppendLine()
+                      .Append(" ---> ");
+                }
+
+                sb.Append(ex.GetType().FullName)
+                  .Append(" (0x")
+                  .Append(ex.HResult.ToString("x", CultureInfo.InvariantCulture))
+                  .Append("): ")
+                  .Append(ex.Message);
+
+                ex = ex.InnerException;
+                nest = true;
+            }
+            while (ex != null);
+
+            return sb.ToString();
```

Comment:
Why can't we just do Exception.ToString here?

Reply:
This style is used in a couple of other places.
I will switch to simply calling `ToString()`.